### DISCUSSION TITLE
Howto push to Gerrit?

### DISCUSSION THREAD

**Discussion by dapc11** [NeogitOrg/neogit#811] [Q&A] [answered]
As the title says, how does one accomplish a push to Gerrit? A push to a "regular" repo, e.g. Github-based, would be something like:
`neogit -> P(ush) -> p(ushRemote)`

Git cmdline equivalent:
`git push`

And that's all to it. But with Gerrit I'm struggling to push, as that looks a bit different to the regular push:
`git push origin HEAD:refs/for/<branch>`

When using Magit I ended up writing my own transient prefix:
`(defun magit-push-to-gerrit ()
  (interactive)
  (magit-git-command "push origin HEAD:refs/for/master" (magit-toplevel)))
`

Is there a way to do this in Neogit?

Thanks in advance.

**Comment by CKolkey** [comment]
Hmm. I think you _could_ do something like:

```lua
local original_fn = require("neogit.lib.git.push").push_interactive

local wrapper_fn = function(remote, branch, args)
  original_fn(remote, "HEAD:refs/for/" .. branch, args)
end

require("neogit.lib.git.push").push_interactive = wrapper_fn
```

If that needs to be changed on a per-project basis, you could consider setting an ENV variable with something like direnv for the project and reading that in the wrapper function.

**Reply by dapc11** [reply]
Thanks for the quick answer. Just a follow-up question. 
Perhaps it's just my lack of knowledge. But this makes all pushes go towards `HEAD:refs/for/" .. branch` right?

Is there a way to add an additional target to the interactive Push. Similar to `e elsewhere`, but instead perhaps: `g gerrit`. Which makes it Neogit -> P -> g

**Reply by CKolkey** [reply, marked as answer]
If you make a fork, sure. Add a line in `lua/neogit/popups/push/init.lua` for a new action, and `action.lua` in that folder for the function. (or just do it inline, doesn't matter). 

But there's no built-in mechanism to do custom popup commands (at this time).

If you have a repo that exclusively uses gerrit, you could do something like `vim.loop.os_getenv("USE_GERRIT")` to determine if you should modify the branch argument being passed or not.

**Reply by dapc11** [reply]
Thanks for the pointer @CKolkey! The fork option is fine for me. Are there plans for custom popup commands?

**Reply by CKolkey** [reply]
Not really.. But I'd accept a PR. I think you'd have to make an interface in the config like:

```lua
{
  NeogitPushPopup = {
    { key = "g", name = "gerrit", action = function(popup) ... end }
  }
}
```

Then apply that somehow in the builder.
